src/shred_collector/repair_service.zig

```diff
@@ -1,101 +1,258 @@
+            );
         }
     }
 
-    fn initialSnapshotRepair(self: *Self) !?AddressedRepairRequest {
-        if (self.slot_to_request == null) return null;
-        const request: RepairRequest = .{ .HighestShred = .{ self.slot_to_request.?, 0 } };
-        const maybe_peer = try self.peer_provider.getRandomPeer(self.slot_to_request.?);
+    const MAX_SHRED_REPAIRS = 1000;
+    const MIN_HIGHEST_REPAIRS = 10;
+    const MAX_HIGHEST_REPAIRS = 200;
 
-        if (maybe_peer) |peer| return .{
-            .request = request,
-            .recipient = peer.pubkey,
-            .recipient_addr = peer.serve_repair_socket,
-        } else {
-            return null;
+    fn getRepairs(self: *Self) !ArrayList(RepairRequest) {
+        var repairs = ArrayList(RepairRequest).init(self.allocator);
+        if (!try self.shred_tracker.identifyMissing(&self.report)) {
+            return repairs;
+        }
+        var individual_count: usize = 0;
+        var highest_count: usize = 0;
+        var slot: Slot = 0;
+
+        var num_highest_repairs: usize = MIN_HIGHEST_REPAIRS;
+        if (self.last_big_request_timestamp_ms + 5_000 < std.time.milliTimestamp()) {
+            self.last_big_request_timestamp_ms = std.time.milliTimestamp();
+            num_highest_repairs = MAX_HIGHEST_REPAIRS;
+        }
+
+        for (self.report.items()) |*report| outer: {
+            slot = report.slot;
+            for (report.missing_shreds.items) |shred_window| {
+                if (shred_window.end) |end| {
+                    for (shred_window.start..end) |i| {
+                        individual_count += 1;
+                        try repairs.append(.{ .Shred = .{ slot, i } });
+                        if (individual_count > MAX_SHRED_REPAIRS) {
+                            break :outer;
+                        }
+                    }
+                }
+            }
+            if (highest_count < num_highest_repairs) {
+                highest_count += 1;
+                try repairs.append(.{ .HighestShred = .{ slot, 0 } });
+            }
         }
+
+        if (highest_count < num_highest_repairs) {
+            for (slot..slot + num_highest_repairs - highest_count) |s| {
+                try repairs.append(.{ .HighestShred = .{ s, 0 } });
+            }
+        }
+        return repairs;
+    }
+
+    fn assignRequestsToPeers(
+        self: *Self,
+        requests: []const RepairRequest,
+    ) !ArrayList(AddressedRepairRequest) {
+        var addressed = ArrayList(AddressedRepairRequest).init(self.allocator);
+        for (requests) |request| {
+            if (try self.peer_provider.getRandomPeer(request.slot())) |peer| {
+                try addressed.append(.{
+                    .request = request,
+                    .recipient = peer.pubkey,
+                    .recipient_addr = peer.serve_repair_socket,
+                });
+            }
+            // TODO do something if a peer is not found?
```

Comment:
Repair could manage its own loop. I just did it with the service manager because it was going to use the same logic as the service manager. But it would be easy to add a run function here with a loop in it if we want a custom way to handle the loop, and the service manager would only be in charge of handling cases that escape the loop.

Reply:
sounds good - will let you decide whats best on the implementation side of things

Reply:
9ea6d768c08d8664881065cf14c38403edd4fa53